### PR TITLE
Collaborative Editing using WebRTC (only 2 peers supported)

### PR DESCRIPTION
**Pull request by gziolo**
Implements: #1930.

Opened in place of #1877 to move development inside the upstream repository.

collab editing

The PR is experimental. Lacks some changes but a starting point.
Please note this is incomplete and is currently being updated.

Detailed Description:
1. We create a Cediting instance inside middleware.

2. Middleware is used to listen to all the actions being dispatched so that we can send those to P2P data channel. Few states are filtered in middleware that we don't send.

3. For peer colors, we use a different action which is always dispatched when you select a block shown on the basis of peerID which is unique to each peer.

4. Locking also happens on the basis of borders. If border for collaboration is visible on one peer side it is locked.

The module used is: https://github.com/abhishekgahlot/gutenberg-rtc
How GRTC works: https://github.com/abhishekgahlot/gutenberg-rtc/blob/master/DESIGN.md

### TODO tasks

- [ ] Serialization of data so it can be transferred to other peers (https://github.com/WordPress/gutenberg/pull/1877#discussion_r129262135). 
- [x] Add proper filtering for the Redux actions that are sent to peers.
- [ ] Fix the edge case for 2 peers if one takes a longer time to connect can result in both peers in the unstable state.
- [ ] Ability to queue data so that when another peer/user comes they can see the changes already done by another peer/user.
- [ ] Optimization for XHR Polling.
- [ ] Better UI/UX for users when they are connected (https://github.com/WordPress/gutenberg/pull/1877#discussion_r142806569).
- [x] Locking of blocks is done by CSS so navigating by keyboard events is still possible. 
- [ ] Fix the case when a collaborator leaves, block remains locked.
- [ ] We don't check user permissions for REST API calls.
- [x] Add missing `@since` in PHP file ( https://github.com/WordPress/gutenberg/pull/1877#discussion_r141022796).
- [x] Extract `withCoediting` HOC to wrap the visual-block component (https://github.com/WordPress/gutenberg/pull/1877#discussion_r141021656).
- [ ] Move Coediting toogle to the ellipsis menu (https://github.com/WordPress/gutenberg/pull/1877#discussion_r141022256).
- [ ] Add support for multi-block select. Make sure frozen blocks are skipped.
- [ ] Add editor's lock when 2 peers collaborate or collaboration disabled.
- [ ] Create a draft before collaboration gets started to avoid duplicate posts/pages
- [ ] Navigating between block with arrows doesn't freeze the active block.

### Blockers

- [ ] [Editable: Treat value as HTML string, eliminate children source ](https://github.com/WordPress/gutenberg/issues/2750)

### Improvements

- [x] [Editable: Invoke onChange on every input](https://github.com/WordPress/gutenberg/issues/2758)

### CODE REVIEW

File: editor/edit-post/sidebar/collaboration/index.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (gziolo) [inline editor/edit-post/sidebar/collaboration/index.js]: Remove this file.

File: coediting/hooks/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+							className={ className }
+							id={ uid }
+							name={ name }
+						/> }
+					</div>
+				</div>
+			);
+		}
+
+		return <OriginalComponent { ...originalProps } />;
```
Comment (gziolo) [inline coediting/hooks/index.js]: It would read better when we flip the condition and return this early.

File: coediting/hooks/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			const blockType = getBlockType( name );
+
+			// Determine whether the block has props to apply to the wrapper.
+			let wrapperProps;
+			if ( blockType.getEditWrapperProps ) {
+				wrapperProps = blockType.getEditWrapperProps( attributes );
+			}
+
+			// Generate a class name for the block's editable form
+			const generatedClassName = hasBlockSupport( blockType, 'className', true ) ?
```
Comment (gziolo) [inline coediting/hooks/index.js]: @aduth @youknowriad can we move it to `EditBlock`? I think it's internal thing which shouldn't be part of this and the original component.

File: coediting/hooks/index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+} from '../../editor/selectors';
+
+const withFrozenMode = ( OriginalComponent, originalProps ) => {
+	const Component = ( { isFrozenByCollaborator, ...props } ) => {
+		if ( isFrozenByCollaborator ) {
+			const { block, collaboratorColor, collaboratorName } = props;
+			const { attributes, name, isValid, uid } = block;
+			const blockType = getBlockType( name );
+
+			// Determine whether the block has props to apply to the wrapper.
```
Comment (gziolo) [inline coediting/hooks/index.js]: @aduth @youknowriad I copied it from the original component. Isn't something that should be part of the `block` object? If I follow properly the flow is as follows:
- in a completely different place use registered block type to create `block`
- her fetch block and its `attributes`
- get block type again to generate `wrapperProps` based on block's `attributes`

Comment (aduth) [reply coediting/hooks/index.js]: >Isn't `wrapperProps ` something that should be part of the `block` object? If I follow properly the flow is as follows:

I'd rather see this move to a block supports. Originally I'd intended for the block list to be unaware of alignment features, but I think it's such a base functionality that it's fine. Might be even better implemented as a hook.

**Comment by gziolo** [comment]
Rebased with master and squashed commits to make future rebases less painful.

File: coediting/hooks/index.js (hunk 3 above)
Comment (gziolo) [reply coediting/hooks/index.js]: Moved in https://github.com/WordPress/gutenberg/pull/4009.

**Comment by gziolo** [comment]
I will rebase it next week. It's still work in progress put aside because it's not planned as a core feature for 5.0.

**Comment by aduth** [comment]
Should we close this until it's given proper focus to be renewed? I can't imagine a rebase at this point will be easy / possible.

**Comment by gziolo** [comment]
Yes, let’s close it 👍

**Pull request closed without merging**